### PR TITLE
[feat] Implement auth and mockbee backend

### PR DESCRIPTION
**Pull request by Hamdrive**
## Overview

Successfully implemented authentication and fully working mock backend to the ecommerce store
Mock backend used is provided by Mockbee
**Note:** Backend logic is persistent until page refresh/rerender

Features:
1. Ability to authenticate user while signup or login
2. Addition of form validation for user credentials input
3. User can now store products in cart or wish list and remains persistent
4.  On logout all local storage data is cleared
5. User data prevails on relogging, cart and wish list data is retrieved from mock backend
6.  Implementation of private routes, which can only be accessed after user authentication
7. Implementation of context and reducer logic to handle authentication and products
8. Entire products, cart, wish list is handled by one global context, thereby easing LOC and maintaining readability

## Deployment

https://deploy-preview-24--fueledup-store.netlify.app/

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -9,9 +9,16 @@
...
+
+    <link
+      rel="stylesheet"
+      href="https://unpkg.com/mockman-js@latest/dist/style.css" />
 
     <!-- PowertrainUI library -->
-    <link rel="stylesheet" href="https://powertrain-ui.netlify.app/styles.css" />
+    <link
+      rel="stylesheet"
+      href="https://powertrain-ui.netlify.app/styles.css" />
```
Comment (ttahm3d) [inline public/index.html:23]: Change to below for better performance

```suggestion
    <link
      rel="preconnect"
      href="https://powertrain-ui.netlify.app" />
    <link
      rel="stylesheet"
      href="https://powertrain-ui.netlify.app/styles.css" />
```

File: src/App.js [hunk 2]
```diff
@@ -1,16 +1,41 @@
...
+        }
+      />
+      <Route
+        path="/wishlist"
+        element={
+          <RequireAuth>
+            <Wishlist />
+          </RequireAuth>
+        }
+      />
```
Comment (ttahm3d) [inline src/App.js]: There is a better way of handling these things. Read more about [outlet](https://reactrouter.com/docs/en/v6/api#outlet)

File: src/components/CartCard.jsx [hunk 3]
```diff
@@ -1,9 +1,31 @@
...
+      handleRemoveFromCart(product);
+    } else {
+      addToWishlist(product);
+      removeFromCart(product._id);
+    }
+  };
+
+  const handleUpdateCartQuantity = (product, type) => {
+    updateCartQuantity(product._id, type);
+  };
```
Comment (ttahm3d) [inline src/components/CartCard.jsx:28]: Nicely done. Even though you are getting the functions from the context, better to have the handlers and handlers calling the functions in the context

File: src/components/PriceSummary.jsx [hunk 4]
```diff
@@ -5,15 +5,15 @@ export function PriceSummary({ products }) {
...
           (Number(item.price) - Number(item.productDiscountPrice))) *
-        Number(item.quantity),
+        Number(item.qty),
       totalPrice:
         (Number(summary.totalPrice) + Number(item.productDiscountPrice)) *
-        Number(item.quantity),
-        quantity: summary.quantity + item.quantity
+        Number(item.qty),
+      quantity: summary.quantity + item.qty,
     }),
```
Comment (ttahm3d) [inline src/components/PriceSummary.jsx]: Move this within reducer

File: src/components/WishlistCard.jsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import { Link, useLocation } from "react-router-dom";
```
Comment (ttahm3d) [inline src/components/WishlistCard.jsx]: Keep this component in the `components` folder only if it is being used in more than one page / view else keep it inside its own page folder

File: src/pages/cart/Cart.jsx [hunk 6]
```diff
@@ -1,31 +1,33 @@
...
-  const { state } = useAuthProducts();
+  const {
+    state: { productsInCart },
+  } = useProduct();
 
   useEffect(() => {
-    document.title = `My Cart (${state["cart"].length}) | FueledUp`;
-  }, [state]);
+    document.title = `My Cart (${productsInCart.length}) | FueledUp`;
+  }, [productsInCart]);
```
Comment (ttahm3d) [inline src/pages/cart/Cart.jsx]: Extract this in useDocumentTitle hook

File: src/utils/server-routes/index.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+export { performUserLogin } from "./server-routes";
+export { performGetCategories } from "./server-routes";
+export { performGetProducts } from "./server-routes";
+export { performAddToCart } from "./server-routes";
+export { performGetCart } from "./server-routes";
+export { performUpdateCartQuantity } from "./server-routes";
+export { performRemoveFromCart } from "./server-routes";
+export { performAddToWishlist } from "./server-routes";
+export { performGetWishlist } from "./server-routes";
+export { performRemoveFromWishlist } from "./server-routes";
```
Comment (ttahm3d) [inline src/utils/server-routes/index.js]:
```suggestion
export { performUserSignup, performUserLogin, performGetCategories, performGetProducts , performAddToCart , performGetCart  , performUpdateCartQuantity ,performRemoveFromCart , performAddToWishlist, performGetWishlist    ,performRemoveFromWishlist } from "./server-routes";
```

**Review by ttahm3d — COMMENTED** [review]
Nice work. Liked the way you are destructuring things and using things from the state.

Have provided few suggestions. Please take a look.

**Review by Ronak-Mutha — APPROVED** [review]
Good use of re-usable components and folder structure. Try to take comments by Tahir into consideration after submission. Good to merge.

File: public/index.html (hunk 1 above)
Comment (Hamdrive) [reply public/index.html:23]: Did not know about preconnect, got to learn tis advantages and how it works.
Added the changes!

File: src/App.js (hunk 2 above)
Comment (Hamdrive) [reply src/App.js]: Learnt what Outlet does and how to implement it, thanks!

File: src/pages/cart/Cart.jsx (hunk 6 above)
Comment (Hamdrive) [reply src/pages/cart/Cart.jsx]: implemented as suggested, thanks!

File: src/utils/server-routes/index.js (hunk 7 above)
Comment (Hamdrive) [reply src/utils/server-routes/index.js]: Did not know of the implications of having multiple exports, got to learn a new concept!

File: src/components/WishlistCard.jsx (hunk 5 above)
Comment (Hamdrive) [reply src/components/WishlistCard.jsx]: Implemented the suggestion

File: src/components/PriceSummary.jsx (hunk 4 above)
Comment (Hamdrive) [reply src/components/PriceSummary.jsx]: converted to reducer function

File: src/components/CartCard.jsx (hunk 3 above)
Comment (Hamdrive) [reply src/components/CartCard.jsx:28]: Noted will try to implement the hierarchy of calls better

**Comment by Hamdrive** [comment]
Successfully got PR reviewed by two team members, changes suggested in review have been implemented and deployment is functional.
Thus this PR can be closed and merged

**Pull request merged**